Add GraalVM Native Image and Truffle compilation sections for GraalPy guides (#1563)
micronaut-projects/micronaut-guides: src/docs/common/snippets/common-graalpy-graalvm.adoc

:exclude-for-languages:groovy

== Python Runtime Compilation

> With GraalVM JDK, the Python code executed using the GraalPy engine will be just-in-time compiled.

Reviewer: ```suggestion
When using the GraalVM JDK, Python code executed with the GraalPy engine is just-in-time compiled.
```

> With other JDKs, the Python code will be interpreted, which negatively impacts its performance.

Reviewer: ```suggestion
With other JDKs, the Python code is interpreted, resulting in reduced performance.
```
I replaced "which negatively impacts its performance" with "resulting in reduced performance"

The easiest way to install https://www.graalvm.org[GraalVM] on Linux or Mac is to use https://sdkman.io/[SDKMan.io].

[source, bash]
.Java 23
----

> sdk install java 23-graal

Reviewer: 23-graal won't work with Gradle afaik

Reviewer: I assume this guide will be only generated for Maven?

Author: Yes, correct

----

For installation on Windows, or for a manual installation on Linux or Mac, see the https://www.graalvm.org/latest/docs/getting-started/[GraalVM Getting Started] documentation.

The previous command installs Oracle GraalVM, which is free to use in production and free to redistribute, at no cost, under the https://www.oracle.com/downloads/licenses/graal-free-license.html[GraalVM Free Terms and Conditions].

Alternatively, you can use the https://github.com/graalvm/graalvm-ce-builds/releases/[GraalVM Community Edition]:

[source, bash]
.Java 23
----
sdk install java 23-graalce
----

=== GraalVM JDK Compatibility

> In order to support runtime compilation, one must use versions of GraalPy and the Polyglot API

Reviewer: ```suggestion
To enable runtime compilation, you must use versions of GraalPy and the Polyglot API dependencies that are compatible with the specified GraalVM JDK version. If you see errors like the following:
```

dependencies compatible with given GraalVM JDK version. If you see errors like this:

[source, bash]
----
Your Java runtime '23.0.1+11-jvmci-b01' with compiler version '24.1.1' is incompatible with polyglot version '24.1.0'.
----

You need to override the versions of the following dependencies according to the error message:

[source, text]
----
org.graalvm.polyglot:polyglot
org.graalvm.python:python
org.graalvm.python:python-embedding
org.graalvm.python:graalpy-maven-plugin
----

== Generate a Micronaut Application Native Executable with GraalVM

> Starting from the GraalVM for JDK 23, Micronaut applications using GraalPy, or any Truffle language in general,

Reviewer: ```suggestion
Starting with GraalVM for JDK 23, Micronaut applications that use GraalPy or any Truffle language can be compiled into native executables using GraalVM Native Image.
```

can be compiled to native executable using GraalVM Native Image.

Compiling Micronaut applications ahead of time with GraalVM significantly improves startup time and reduces
the memory footprint of JVM-based applications.

> === Native Executable metadata

Reviewer: ```
=== Required Metadata
```
I would suggest to change, we don't use the phrase like "Native Executable metadata", but "Reachability Metadata" or "Proxy Configuration"

Author: Changing to GraalVM Native Image Metadata, sg?

Reviewer: Could be also just "Native Image Configuration"

> The https://www.graalvm.org/[GraalVM] Native Image compilation requires metadata to properly run code that uses https://www.graalvm.org/latest/reference-manual/native-image/metadata/#dynamic-proxy[dynamic proxies].

Reviewer: ```suggestion
GraalVM Native Image compilation requires metadata to correctly run code that uses https://www.graalvm.org/latest/reference-manual/native-image/metadata/#dynamic-proxy[dynamic proxies].
```

Create a proxy configuration file:
resource:META-INF/native-image/proxy-config.json[]

=== Native Executable Generation

To generate a native executable using Maven, run:

[source, bash]
----
./mvnw package -Dpackaging=native-image
----

> The native executable is created in the `target` directory and can be run with `target/micronautguide`.

Reviewer: ```suggestion
The native executable is created in the _target_ directory and can be run with `./target/micronautguide`.
```

=== GraalVM JDK Compatibility

To enable runtime compilation, you must use versions of GraalPy and the Polyglot API dependencies that are compatible with the specified GraalVM JDK version. If you see errors like the following:

> dependencies compatible with given GraalVM JDK version. If you see errors like this:

Reviewer: A leftover

== Generate a Micronaut Application Native Executable with GraalVM

Starting with GraalVM for JDK 23, Micronaut applications that use GraalPy or any Truffle language can be compiled into native executables using GraalVM Native Image.

> can be compiled to native executable using GraalVM Native Image.

Reviewer: A leftover
